backend/internal/handlers/spahandler.go

```diff
@@ -0,0 +1,57 @@
+// As suggested by https://github.com/gorilla/mux#serving-single-page-applications
+
+package handlers
+
+import (
+	"net/http"
+	"os"
+	"path/filepath"
+)
+
+// SpaHandler implements the http.Handler interface, so we can use it
+// to respond to HTTP requests. The path to the static directory and
+// path to the index file within that static directory are used to
+// serve the SPA in the given static directory.
+type SpaHandler struct {
+	staticPath string
+	indexPath  string
+}
+
+// NewSpaHandler creates and returns a SpaHandler
+func NewSpaHandler(staticPath, indexPath string) SpaHandler {
+	return SpaHandler{staticPath: staticPath, indexPath: indexPath}
+}
+
+// ServeHTTP inspects the URL path to locate a file within the static dir
+// on the SPA handler. If a file is found, it will be served. If not, the
+// file located at the index path on the SPA handler will be served. This
+// is suitable behavior for serving an SPA (single page application).
+func (h SpaHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	// get the absolute path to prevent directory traversal
+	path, err := filepath.Abs(r.URL.Path)
+	if err != nil {
+		// if we failed to get the absolute path respond with a 400 bad request
+		// and stop
+		http.Error(w, err.Error(), http.StatusBadRequest)
+		return
+	}
+
+	// prepend the path with the path to the static directory
+	path = filepath.Join(h.staticPath, path)
+
+	// check whether a file exists at the given path
+	_, err = os.Stat(path)
+	if os.IsNotExist(err) {
+		// file does not exist, serve index.html
+		http.ServeFile(w, r, filepath.Join(h.staticPath, h.indexPath))
+		return
```

Comment:
Shouldn't you just return a 404 and only `index.html` if it's requested? @alex-kovoy thoughts?

Reply:
My understanding is that it's written that way in order to support browser-routing (in this case, `react-router`). With this configuration, if a user opens a new tab and navigates directly to a browser-routed route like https://localhost:8000/dashboard, the server will fall back to the index.html and just serve them the react app, which will then recognize the route and load the Dashboard component.

A downside to this is that if the user navigates directly to https://localhost:8000/dne, they'll be shown a blank page. A way that I've dealt with that issue in the past is by creating a "404 not found" react component and route and then redirecting any non-existent routes to it (though the server would in that case still return index.html/200).

Reply:
@alex-kovoy @ibeckermayer I see, thanks for the explanation.

Reply:
This seems to create a tight coupling between frontend and backend logic.
But I don't really write web applications these days, so maybe I just need to get on with the times.